packages/core/src/classes/query-string.class.ts

```diff
@@ -71,3 +74,8 @@ export class QueryString {
     return _res;
   }
 }
+
+function assertAllowedPropertyName(name: string): asserts name {
+  if (PROPERTY_NAME_DENY_LIST.indexOf(name) >= 0)
+    throw new Error(`Property name "${name}" is not allowed`);
+}
```

Comment:
TIL about https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions!

In our case, you're doing `assertAllowedPropertyName(part)` and then asserting just that string value. From reading the TS docs, shouldn't it be closer to:

```ts
assertAllowedPropertyName(!PROPERTY_NAME_DENY_LIST.includes(name), name)

function assertAllowedPropertyName(condition: boolean, name: string): asserts condition {
  if (!condition)
    throw new Error(`Property name "${name}" is not allowed`);
}
```

where we pass the condition as the argument? I've never used these assertions so not fully sure how they're meant to work!

Reply:
Not quite, the assertion function can be used to evaluate an expression, and inform the compiler that that expression must be true for the rest of the evaluation - but it can also be used to narrow a type.

In the narrowing example it's saying that a type can be cast what ever the assert is specifying with `is`, if `is` is missing then it's not narrowing but more of a developer hint that this might throw.

If typescript's type system allowed what I would actually be specifying is `asserts name is not "__proto__" | "prototype"`, but it doesn't.

Reply:
there's some other examples here https://www.typescriptlang.org/play#example/assertion-functions